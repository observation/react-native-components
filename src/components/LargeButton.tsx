import React from 'react'
import { StyleProp, StyleSheet, Text, TouchableOpacity, View, ViewStyle } from 'react-native'

import { Icon } from './Icon'
import { IconName } from '../lib/Icons'
import * as LargeButtonStyles from '../lib/LargeButtonStyles'
import { LargeButtonStyle } from '../lib/LargeButtonStyles'
import Log from '../lib/Log'
import appTextStyle from '../styles/text'
import theme from '../styles/theme'

type LargeButtonProps = {
  title: string
  onPress?: () => void
  onPressIn?: () => void
  iconName?: IconName
  disabled?: boolean
  secondary?: boolean
  danger?: boolean
  style?: StyleProp<ViewStyle>
  testID?: string
}

const getStyle = (secondary?: boolean, disabled?: boolean, danger?: boolean): LargeButtonStyle => {
  Log.trace('LargeButton:getStyle')

  const enabled = !disabled
  const primary = !secondary

  switch (true) {
    case primary && enabled && danger:
      return LargeButtonStyles.primaryDanger
    case primary && enabled && !danger:
      return LargeButtonStyles.primary
    case primary && disabled && !danger:
      return LargeButtonStyles.primaryDisabled
    case secondary && enabled && danger:
      return LargeButtonStyles.secondaryDanger
    case secondary && enabled && !danger:
      return LargeButtonStyles.secondary
    case secondary && disabled && !danger:
      return LargeButtonStyles.secondaryDisabled

    default:
      return LargeButtonStyles.primary
  }
}

const LargeButton = ({
  secondary,
  disabled,
  danger,
  style,
  title,
  iconName,
  onPress,
  onPressIn,
  testID = 'touchable-opacity',
}: LargeButtonProps) => {
  const { textStyle, buttonStyle, iconColor } = getStyle(secondary, disabled, danger)

  return (
    <TouchableOpacity
      testID={testID}
      style={[styles.container, style, buttonStyle]}
      disabled={disabled}
      onPress={onPress}
      onPressIn={onPressIn}
      activeOpacity={0.5}
    >
      <View style={styles.titleContainer}>
        {iconName && (
          <View style={styles.iconContainerStyle}>
            <Icon name={iconName} size={theme.icon.size.large} color={iconColor} />
          </View>
        )}
        <Text style={[styles.title, textStyle]}>{title}</Text>
      </View>
    </TouchableOpacity>
  )
}

export default LargeButton
export type { LargeButtonProps }

const styles = StyleSheet.create({
  container: {
    ...theme.rounded,
    margin: theme.margin.common,
    height: 32,
    justifyContent: 'center',
    alignItems: 'center',
  },
  title: {
    textAlignVertical: 'center',
    ...appTextStyle.lead,
  },
  titleContainer: {
    marginHorizontal: theme.margin.common,
    flexDirection: 'row',
  },
  iconContainerStyle: {
    justifyContent: 'center',
    paddingRight: theme.margin.half,
  },
})
