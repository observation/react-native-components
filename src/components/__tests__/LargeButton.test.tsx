import React from 'react'

import { render, fireEvent } from '@testing-library/react-native'

import LargeButton from '../LargeButton'

const onPress = jest.fn()
const onPressIn = jest.fn()

describe('LargeButton', () => {
  test('Rendering, enabled, primary', () => {
    const { toJSON } = render(<LargeButton title="Press me" onPress={onPress} style={{ flex: 1 }} />)
    expect(toJSON()).toMatchSnapshot()
  })

  test('Rendering, enabled, secondary', () => {
    const { toJSON } = render(<LargeButton secondary title="Press me" onPress={onPress} style={{ flex: 1 }} />)
    expect(toJSON()).toMatchSnapshot()
  })

  test('Rendering, disabled, primary', () => {
    const { toJSON } = render(<LargeButton disabled title="Press me" onPress={onPress} style={{ flex: 1 }} />)
    expect(toJSON()).toMatchSnapshot()
  })

  test('Rendering, disabled, secondary', () => {
    const { toJSON } = render(<LargeButton disabled secondary title="Press me" onPress={onPress} style={{ flex: 1 }} />)
    expect(toJSON()).toMatchSnapshot()
  })

  test('Rendering, secondary, danger', () => {
    const { toJSON } = render(<LargeButton secondary danger title="Press me" onPress={onPress} style={{ flex: 1 }} />)
    expect(toJSON()).toMatchSnapshot()
  })

  test('Rendering, secondary, danger, disabled => defaults to primary', () => {
    const { toJSON } = render(
      <LargeButton secondary disabled danger title="Press me" onPress={onPress} style={{ flex: 1 }} />,
    )
    expect(toJSON()).toMatchSnapshot()
  })

  test('Rendering, with icon', () => {
    const { toJSON } = render(
      <LargeButton iconName="info-circle" title="Press me" onPress={onPress} style={{ flex: 1 }} />,
    )
    expect(toJSON()).toMatchSnapshot()
  })

  test('Click', async () => {
    const { getByTestId } = render(<LargeButton title="Press me" onPress={onPress} style={{ flex: 1 }} />)
    await fireEvent.press(getByTestId('touchable-opacity'))
    expect(onPress).toBeCalled()
  })

  test('onPressIn', async () => {
    const { getByTestId } = render(
      <LargeButton title="Press me" onPress={onPress} onPressIn={onPressIn} style={{ flex: 1 }} />,
    )
    await fireEvent(getByTestId('touchable-opacity'), 'pressIn')
    expect(onPressIn).toBeCalled()
  })
})
