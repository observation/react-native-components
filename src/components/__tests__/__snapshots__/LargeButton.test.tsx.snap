// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LargeButton Rendering, disabled, primary 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#E6E6E6",
      "borderRadius": 4,
      "flex": 1,
      "height": 32,
      "justifyContent": "center",
      "margin": 16,
      "opacity": 1,
      "overflow": "hidden",
    }
  }
  testID="touchable-opacity"
>
  <View
    style={
      {
        "flexDirection": "row",
        "marginHorizontal": 16,
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#212121",
            "fontFamily": "Ubuntu",
            "fontSize": 14,
            "fontStyle": "normal",
            "fontWeight": "bold",
            "lineHeight": 20,
            "textAlignVertical": "center",
          },
          {
            "color": "#FFFFFF",
          },
        ]
      }
    >
      Press me
    </Text>
  </View>
</View>
`;

exports[`LargeButton Rendering, disabled, secondary 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "borderColor": "#E6E6E6",
      "borderRadius": 4,
      "borderWidth": 2,
      "flex": 1,
      "height": 32,
      "justifyContent": "center",
      "margin": 16,
      "opacity": 1,
      "overflow": "hidden",
    }
  }
  testID="touchable-opacity"
>
  <View
    style={
      {
        "flexDirection": "row",
        "marginHorizontal": 16,
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#212121",
            "fontFamily": "Ubuntu",
            "fontSize": 14,
            "fontStyle": "normal",
            "fontWeight": "bold",
            "lineHeight": 20,
            "textAlignVertical": "center",
          },
          {
            "color": "#E6E6E6",
          },
        ]
      }
    >
      Press me
    </Text>
  </View>
</View>
`;

exports[`LargeButton Rendering, enabled, primary 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#0066B1",
      "borderRadius": 4,
      "flex": 1,
      "height": 32,
      "justifyContent": "center",
      "margin": 16,
      "opacity": 1,
      "overflow": "hidden",
    }
  }
  testID="touchable-opacity"
>
  <View
    style={
      {
        "flexDirection": "row",
        "marginHorizontal": 16,
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#212121",
            "fontFamily": "Ubuntu",
            "fontSize": 14,
            "fontStyle": "normal",
            "fontWeight": "bold",
            "lineHeight": 20,
            "textAlignVertical": "center",
          },
          {
            "color": "#FFFFFF",
          },
        ]
      }
    >
      Press me
    </Text>
  </View>
</View>
`;

exports[`LargeButton Rendering, enabled, secondary 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "borderColor": "#0066B1",
      "borderRadius": 4,
      "borderWidth": 2,
      "flex": 1,
      "height": 32,
      "justifyContent": "center",
      "margin": 16,
      "opacity": 1,
      "overflow": "hidden",
    }
  }
  testID="touchable-opacity"
>
  <View
    style={
      {
        "flexDirection": "row",
        "marginHorizontal": 16,
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#212121",
            "fontFamily": "Ubuntu",
            "fontSize": 14,
            "fontStyle": "normal",
            "fontWeight": "bold",
            "lineHeight": 20,
            "textAlignVertical": "center",
          },
          {
            "color": "#0066B1",
          },
        ]
      }
    >
      Press me
    </Text>
  </View>
</View>
`;

exports[`LargeButton Rendering, primary, danger 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#EA554B",
      "borderRadius": 4,
      "flex": 1,
      "height": 32,
      "justifyContent": "center",
      "margin": 16,
      "opacity": 1,
      "overflow": "hidden",
    }
  }
  testID="touchable-opacity"
>
  <View
    style={
      {
        "flexDirection": "row",
        "marginHorizontal": 16,
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#212121",
            "fontFamily": "Ubuntu",
            "fontSize": 14,
            "fontStyle": "normal",
            "fontWeight": "bold",
            "lineHeight": 20,
            "textAlignVertical": "center",
          },
          {
            "color": "#FFFFFF",
          },
        ]
      }
    >
      Press me
    </Text>
  </View>
</View>
`;

exports[`LargeButton Rendering, secondary, danger 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "borderColor": "#EA554B",
      "borderRadius": 4,
      "borderWidth": 2,
      "flex": 1,
      "height": 32,
      "justifyContent": "center",
      "margin": 16,
      "opacity": 1,
      "overflow": "hidden",
    }
  }
  testID="touchable-opacity"
>
  <View
    style={
      {
        "flexDirection": "row",
        "marginHorizontal": 16,
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#212121",
            "fontFamily": "Ubuntu",
            "fontSize": 14,
            "fontStyle": "normal",
            "fontWeight": "bold",
            "lineHeight": 20,
            "textAlignVertical": "center",
          },
          {
            "color": "#EA554B",
          },
        ]
      }
    >
      Press me
    </Text>
  </View>
</View>
`;

exports[`LargeButton Rendering, secondary, danger, disabled => defaults to primary 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#0066B1",
      "borderRadius": 4,
      "flex": 1,
      "height": 32,
      "justifyContent": "center",
      "margin": 16,
      "opacity": 1,
      "overflow": "hidden",
    }
  }
  testID="touchable-opacity"
>
  <View
    style={
      {
        "flexDirection": "row",
        "marginHorizontal": 16,
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#212121",
            "fontFamily": "Ubuntu",
            "fontSize": 14,
            "fontStyle": "normal",
            "fontWeight": "bold",
            "lineHeight": 20,
            "textAlignVertical": "center",
          },
          {
            "color": "#FFFFFF",
          },
        ]
      }
    >
      Press me
    </Text>
  </View>
</View>
`;

exports[`LargeButton Rendering, with icon 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#0066B1",
      "borderRadius": 4,
      "flex": 1,
      "height": 32,
      "justifyContent": "center",
      "margin": 16,
      "opacity": 1,
      "overflow": "hidden",
    }
  }
  testID="touchable-opacity"
>
  <View
    style={
      {
        "flexDirection": "row",
        "marginHorizontal": 16,
      }
    }
  >
    <View
      style={
        {
          "justifyContent": "center",
          "paddingRight": 8,
        }
      }
    >
      <Icon
        color="#FFFFFF"
        name="circle-info"
        size={16}
        type="light"
      />
    </View>
    <Text
      style={
        [
          {
            "color": "#212121",
            "fontFamily": "Ubuntu",
            "fontSize": 14,
            "fontStyle": "normal",
            "fontWeight": "bold",
            "lineHeight": 20,
            "textAlignVertical": "center",
          },
          {
            "color": "#FFFFFF",
          },
        ]
      }
    >
      Press me
    </Text>
  </View>
</View>
`;
